File: frontend/lib/types/scryfall.ts
/**
 * Source: https://scryfall.com/docs/api/cards
 */

export interface ScryfallCard {
  object: "card";
  _id: string;
  id: string;
  oracle_id: string;
  multiverse_ids: number[];
  resource_id?: string;
  mtgo_id?: number;
  mtgo_foil_id?: number;
  arena_id?: number;
  tcgplayer_id?: number;
  cardmarket_id?: number;
  name: string;
  lang: string;
  released_at: string;
  uri: string;
  scryfall_uri: string;
  layout: string;
  highres_image: boolean;
  image_status: string;
  image_uris: {
    small: string;
    normal: string;
    large: string;
    png: string;
    art_crop: string;
    border_crop: string;
  } | null;
  card_faces?: ScryfallCardFace[];
  mana_cost: string;
  cmc: number;
  type_line: string;
  oracle_text: string;
  power?: string;
  toughness?: string;
  colors: ScryfallCardColor[];
  color_identity: ScryfallCardColor[];
  keywords: string[];
  produced_mana?: string[];
  legalities: ScryfallCardLegalities;
  games: string[];
  reserved: boolean;
  game_changer: boolean;
  foil: boolean;
  nonfoil: boolean;
  finishes: string[];
  oversized: boolean;
  promo: boolean;
  reprint: boolean;
  variation: boolean;
  set_id: string;
  set: string;
  set_name: string;
  set_type: string;
  set_uri: string;
  set_search_uri: string;
  scryfall_set_uri: string;
  rulings_uri: string;
  prints_search_uri: string;
  collector_number: string;
  digital: boolean;
  rarity: string;
  flavor_text?: string;
  card_back_id: string;
  artist: string;
  artist_ids: string[];
  illustration_id: string;
  border_color: string;
  frame: string;
  frame_effects?: string[];
  full_art: boolean;
  textless: boolean;
  booster: boolean;
  story_spotlight: boolean;
  edhrec_rank?: number;
  penny_rank?: number;
  prices: ScryfallCardPrices;
  related_uris: ScryfallCardRelatedUris;
  purchase_uris: ScryfallCardPurchaseUris;
}

export interface MongoDBScryfallCard extends ScryfallCard {
  _id: string;
}

export interface ScryfallCardLegalities {
  [format: string]: "legal" | "not_legal" | "restricted" | "banned";
}

export interface ScryfallCardFace {
  name: string;
  mana_cost: number;
  type_line: string;
  oracle_text?: string;
  colors: ScryfallCardColor;
  image_uris?: {
    small: string;
    normal: string;
    large: string;
    png: string;
    art_crop: string;
    border_crop: string;
  };
}

export interface ScryfallCardPrices {
  usd: string | null;
  usd_foil: string | null;
  usd_etched: string | null;
  eur: string | null;
  eur_foil: string | null;
  tix: string | null;
}

export interface ScryfallCardRelatedUris {
  gatherer?: string;
  tcgplayer_infinite_articles?: string;
  tcgplayer_infinite_decks?: string;
  edhrec?: string;
  [key: string]: string | undefined;
}

export interface ScryfallCardPurchaseUris {
  tcgplayer?: string;
  cardmarket?: string;
  cardhoarder?: string;
  [key: string]: string | undefined;
}

export type ScryfallCardColor = "W" | "R" | "B" | "U" | "G" | null;

export interface ScryfallCardApiResponse {
  object: "list";
  total_cards: number;
  has_more: boolean;
  next_page?: string;
  data: ScryfallCard[];
}

File: backend/app/models/scryfall.py
from typing import Literal

from pydantic import BaseModel, ConfigDict

ScryfallCardLegality = Literal["legal", "not_legal", "restricted", "banned"]
ScryfallCardColor = Literal["W", "R", "B", "U", "G"] | None


class ScryfallImageUris(BaseModel):
    small: str
    normal: str
    large: str
    png: str
    art_crop: str
    border_crop: str


class ScryfallCardFace(BaseModel):
    name: str
    mana_cost: str
    type_line: str | None = None
    oracle_text: str | None = None
    colors: list[ScryfallCardColor] | ScryfallCardColor | None = None
    image_uris: ScryfallImageUris | None = None


class ScryfallCardPrices(BaseModel):
    usd: str | None = None
    usd_foil: str | None = None
    usd_etched: str | None = None
    eur: str | None = None
    eur_foil: str | None = None
    tix: str | None = None


class ScryfallCardRelatedUris(BaseModel):
    model_config = ConfigDict(extra="allow")

    gatherer: str | None = None
    tcgplayer_infinite_articles: str | None = None
    tcgplayer_infinite_decks: str | None = None
    edhrec: str | None = None


class ScryfallCardPurchaseUris(BaseModel):
    model_config = ConfigDict(extra="allow")

    tcgplayer: str | None = None
    cardmarket: str | None = None
    cardhoarder: str | None = None


class ScryfallCardBase(BaseModel):
    model_config = ConfigDict(extra="ignore", populate_by_name=True)
    object: Literal["card"]
    oracle_id: str
    multiverse_ids: list[int]
    resource_id: str | None = None
    mtgo_id: int | None = None
    mtgo_foil_id: int | None = None
    arena_id: int | None = None
    tcgplayer_id: int | None = None
    cardmarket_id: int | None = None
    name: str
    lang: str
    released_at: str
    uri: str
    scryfall_uri: str
    layout: str
    highres_image: bool
    image_status: str
    image_uris: ScryfallImageUris | None = None
    card_faces: list[ScryfallCardFace] | None = None
    mana_cost: str | None = None
    cmc: int | float
    type_line: str
    oracle_text: str | None = None
    power: str | None = None
    toughness: str | None = None
    colors: list[ScryfallCardColor] = []
    color_identity: list[ScryfallCardColor]
    keywords: list[str]
    produced_mana: list[str] | None = None
    legalities: dict[str, ScryfallCardLegality]
    games: list[str]
    reserved: bool
    game_changer: bool
    foil: bool
    nonfoil: bool
    finishes: list[str]
    oversized: bool
    promo: bool
    reprint: bool
    variation: bool
    set_id: str
    set: str
    set_name: str
    set_type: str
    set_uri: str
    set_search_uri: str
    scryfall_set_uri: str
    rulings_uri: str
    prints_search_uri: str
    collector_number: str
    digital: bool
    rarity: str
    flavor_text: str | None = None
    card_back_id: str | None = None
    artist: str
    artist_ids: list[str] | None = None
    illustration_id: str | None = None
    border_color: str
    frame: str
    frame_effects: list[str] | None = None
    full_art: bool
    textless: bool
    booster: bool
    story_spotlight: bool
    edhrec_rank: int | None = None
    penny_rank: int | None = None
    prices: ScryfallCardPrices
    related_uris: ScryfallCardRelatedUris | None = None
    purchase_uris: ScryfallCardPurchaseUris | None = None


class ScryfallCard(ScryfallCardBase):
    id: str


class ScryfallCardApiResponse(BaseModel):
    object: Literal["list"]
    total_cards: int
    has_more: bool
    next_page: str | None = None
    data: list[ScryfallCard]
